tests/input.rs

```diff
@@ -0,0 +1,299 @@
+use bevy::{
+    asset::AssetPlugin,
+    core::CorePlugin,
+    core_pipeline::CorePipelinePlugin,
+    ecs::event::Events,
+    input::InputPlugin,
+    input::{keyboard::KeyboardInput, ButtonState},
+    pbr::PbrPlugin,
+    prelude::{App, Entity},
+    prelude::{Commands, Component, EventReader, KeyCode},
+    render::RenderPlugin,
+    utils::HashSet,
+    window::{ReceivedCharacter, WindowId, WindowPlugin},
+};
+
+#[derive(Component)]
+struct CharComponent(char);
+fn spawn_entity_at_char_event_system(
+    mut commands: Commands,
+    mut char_input_events: EventReader<ReceivedCharacter>,
+) {
+    for event in char_input_events.iter() {
+        commands.spawn().insert(CharComponent(event.char));
+    }
+}
+
+#[derive(Component)]
+struct KeyComponent(KeyboardInput);
+fn spawn_entity_at_keyboard_event_system(
+    mut commands: Commands,
+    mut char_input_events: EventReader<KeyboardInput>,
+) {
+    for event in char_input_events.iter() {
+        commands.spawn().insert(KeyComponent(event.clone()));
+    }
+}
+
+macro_rules! get_entities {
+    ($app: ident, $query: tt) => {{
+        let world = &mut $app.world;
+        let mut query = world.query::<$query>();
+        query.iter(world).collect::<Vec<_>>()
+    }};
+}
+
+macro_rules! send_event {
+    ($app: ident, $t: ty, $event: expr) => {
```

Comment:
Is it possible to get the type from the input without having to repeat the type as argument?